### ISSUE TITLE
Trial running the rechunking tool on Beam's Dask Runner

### ISSUE THREAD

**Issue by mattjbr123** [NERC-CEH/dri_gridded_data#31]

**Comment by metazool** [comment]
Can I ask a bit about where you're running Beam?

Is this on JASMIN, or is it AWS-based? Is there somewhere we can read through the configuration? 

cc @albags

**Comment by mattjbr123** [comment]
Originally the idea is to trial it on JASMIN, specifically LOTUS, their HPC-like cluster, which has dask-gateway for ease of spinning up dask clusters

> Is there somewhere we can read through the configuration?
There will be once I've actually had a go at it!

**Comment by mattjbr123** [comment]
I realise a 'getting started' blurb and some more info might be helpful. 
There are minimal instructions in the [README](https://github.com/NERC-CEH/dri_gridded_data) for creating the necessary environment and running the script that uses Beam. Currently doesn't cover needing access to the data itself, which is on a ['group_workspace'](https://help.jasmin.ac.uk/docs/short-term-project-storage/introduction-to-group-workspaces/) (read 'access controlled disk space') on JASMIN (which I can happilly grant you access to if you go down that route). 
So you might not be able to actually run the script yet but maybe looking at it is still somewhat helpful. 

There is limited configuration needed for Beam itself at this stage of our development - it largely figures stuff out for itself with the pipeline you've given it to run. The only things [we've tweaked so far](https://github.com/NERC-CEH/dri_gridded_data/blob/5ea0ed9c23a766d8ea30e80b823ad1c63b3fd452/scripts/GEAR/convert_GEAR_beam.py#L114) is 'num_workers' when running in parallel, the 'direct_running_mode' to tell it to run in serial or parallel, and 'auto_unique_labels' which allows multiple steps in the pipeline of the same name (e.g. Print, useful for debugging). 
So far we have run it 'locally', which in this case means on a JASMIN VM or PM (physical machine), and on LOTUS, JASMIN's HPC-like cluster, both using the ['Direct Runner'](https://beam.apache.org/documentation/runners/direct/) which is the built-in no-frills generic 'Runner' which means Beam can run anywhere. 

The next step, covered by this here issue, is to move away from the limitations of the Direct Runner and use another, the first to try being the [Dask Runner](https://dask.discourse.group/t/distributed-runner-for-beam/784). What this does is allow a Beam pipeline to be translated into Dask parlance/objects/task-graphs etc. and run it wherever you have a dask cluster, such as the [dask gateway](https://help.jasmin.ac.uk/docs/interactive-computing/dask-gateway/) that JASMIN maintains. This will probably require some more configuration, and is what I'm currently working on.

**Comment by mattjbr123** [comment]
Xarray then would pick up this client by default, without any further config, whereas beam will require the URL to be provided.
Next commit has added a copy of the script adapted for using dask & beam, or at least how I think it should work!

**Comment by mattjbr123** [comment]
I have managed to get the pipeline to start running on the DaskRunner (more details on how below), but it is running into memory problems early on that were not present when using the DirectRunner in parallel. I'm not sure if this is more to do with the newness and lacking of some features of the DaskRunner or whether this is expected behaviour when using Dask on a workflow like this. I have asked on the pangeo discourse for advice! https://discourse.pangeo.io/t/daskrunner-and-pangeo-forge-recipes/4776

JASMIN's dask gateway has a memory limit of 20GB per worker which is not enough to resolve the issue (though when running with Beam's DirectRunner each worker never uses more than ~1GB of memory, so I feel like it _should_ be enough...) One easy thing I can try is to throw more memory at the problem, but running with DaskRunner on a local or SLURM cluster instead to avoid the low memory limit of the gateway. 

I had to make changes to the the DaskRunner code, essentially moving the [config/setup stage](https://help.jasmin.ac.uk/docs/interactive-computing/dask-gateway/#in-the-jasmin-notebooks-service) of the dask cluster inside the DaskRunner code so that it would more easily work with the gateway cluster. This necessesitated some further changes to the DaskRunner code to enable me to pass in the dask config options to it. These changes I would look to formalise/tidy-up and merge into the Beam library via a PR if it's a setup that ends up working well. 
I've included a copy of this in the next commit to keep track of it.

**Comment by mattjbr123** [comment]
Does seem to work running on SLURM, bypassing dask-gateway, at least when I give the workers 50GB each anyway... If it works with 20GB, then this is the way forward, but I think that's unlikely.

**Comment by mattjbr123** [comment]
Some helpful info from @alxmrs : https://discourse.pangeo.io/t/daskrunner-and-pangeo-forge-recipes/4776/3
Appears that the dask graph that is created from the beam pipeline isn't yet optimized, which might be causing the issue.

**Comment by mattjbr123** [comment]
Meanwhile I will have a go at using the SparkRunner, and see if I get anywhere with that... #27
